migrated persistence services from openhab1-addons

kaikreuzer opened the pull request:
Depends on https://github.com/openhab/openhab-core/pull/1387

Signed-off-by: Kai Kreuzer <kai@openhab.org>

ssalonen commented:
PR for DynamoDB: https://github.com/kaikreuzer/openhab2-addons/pull/4

kaikreuzer commented:
@davidgraeff Question to you as the MQTT expert: The MQTT persistence is linked to the MQTT 1.x binding. Is there already any similar functionality in 2.x? Are people using it at all / is it still useful?
Will we need to adapt the persistence service to the new MQTT support?

davidgraeff commented:
> Is there already any similar functionality in 2.x?

No not yet.

> Are people using it at all / is it still useful?

People are not using it currently. There was not a single complain with the big OH 2.4 MQTT introduction that the mqtt1 persistence broke for anyone.

> Will we need to adapt the persistence service to the new MQTT support?

I'm not entirely sure what this persistence is supposed to do to be honest. A 5 lines rule can already push every item change to specific MQTT topics and I want to add OH<-->OH eventbus synchronisation via MQTT as well. I really don't see a reason for this extension?!

kaikreuzer commented:
I agree that this isn't a very good match for a persistence service - so let's drop it here and we can still re-introduce it, if somebody comes up with a good use case for it.

kaikreuzer commented:
@derekadams: Trying to migrate the Sitewhere persistence service, it deemed to me that this code is quite old and might not be working with the new Sitewhere 2.0 service anymore - the used MQTT communication rather seems to be replaced by a REST API, and the referenced tutorial at https://sitewhere.io/integration/openhab.html meanwhile results in a 404. Instead of putting much effort into it, I'll rather remove it here. If you believe that it is worthwhile to update it and keep it available for users, please let us know and help on getting it up-to-date again!

kaikreuzer commented:
Putting this on hold for the moment, see https://github.com/openhab/openhab-core/pull/680#issuecomment-487411016.

derekadams commented:
@kaikreuzer We are planning to do a new persistence service for the SiteWhere 2.0 architecture since some of the interaction patterns have changed. Thanks for taking a look at the existing code. You're correct that we leverage the updated REST APIs to do much of the work now rather than sending the payloads via MQTT. We may support both options, but will know more once we dive back into the persistence service code. Thanks!

kaikreuzer commented:
Thanks @derekadams, I'll let you know once this is merged (after the 2.5 release) and I'll remove the old Sitewhere service, so that we can add the new one then!

JueBag commented:
I observed an issue concerning rrd4j ( filed [here](https://github.com/openhab/openhab1-addons/issues/5854) ). 
.rrd files are created when performing a REST call although the items are persisted.
Not sure if that is an issue or a feature.

kaikreuzer commented:
@openhab/add-ons-maintainers I'd say this is finally ready for review.

A few notes:

I have
- ported dynamodb, influxdb, jdbc, jpa, mongodb, mysql and rrd4j persistence services over from openhab1-addons.
- ported mapdb persistence service from the 2.5.x branch of openhab-addons (where it is useless)
- adapted the services to the latest interface of openhab-core
- added null annotations on the OSGi services
- replaced component xmls by annotations
- embedded all dependencies within the bundles, so that all of them resolve and start up without any ClassNotFoundExceptions.
- fixed all SAT errors on the code
- successfully functionally tested rrd4j with the 3.0 distro
- successfully started the bundles of all services without errors or exceptions in the log

I have NOT:
- ported db4o as it seems to be dead since years and no artifacts are available on Maven Central.
- ported caldav as it has a dependency on the caldav io bundle, which I didn't want to port as caldav is highly complex and https://github.com/openhab/openhab-addons/pull/6453 might be a better working alternative.
- ported cosm as it was renamed to xively and then bought and shut down by Google
- ported exec because I feel that this isn't really relevant and any use case should be doable with the exec binding and some simple rules
- ported gcal as it also required an io bundle that I wasn't clear whether we should port it or embed it. Also not sure myself, if it is at all working as I remember many discussions around it. So something t.b.d.
- ported logging, because I doubt anybody is using that - we have a flexible logging framework as a replacement by now
- ported mqtt as it depends on the old mqtt code and I again think that it's not useful in times of the new MQTT binding.
- ported Sense as it has been discontinued
- ported Sitewhere because of https://github.com/openhab/openhab-addons/pull/5275#issuecomment-487414048
- not refactored any code besides rrd4j, but preferred to keep it as much as possible untouched by the move, so that the original maintainers can take over from there.
- performed functional tests on any other than the rrd4j service. My suggestion would be to nonetheless merge this PR and then have individual maintainers of the different services work on independent PRs to improve/fix the respective services, if needed.

So in summary: Please don't spend much time on reviewing the code itself - this is all just moved, I didn't implement it and I won't have the time&energy to address any details in the code. So please rather help to merge that quickly and then have the individual persistence maintainers actively go after the code.

cpmeister commented on bundles/org.openhab.persistence.jpa/src/main/java/org/openhab/persistence/jpa/internal/JpaPersistenceService.java:
```diff
@@ -0,0 +1,326 @@
+     * @return initialized EntityManagerFactory
+     */
+    protected EntityManagerFactory newEntityManagerFactory() {
+        logger.debug("Creating EntityManagerFactory...");
```
This is more appropriate as trace logging

cpmeister commented on bundles/org.openhab.persistence.mapdb/src/main/java/org/openhab/persistence/mapdb/internal/StateTypeAdapter.java:
```diff
@@ -0,0 +1,70 @@
+ *
+ * @author Martin Kühl - Initial contribution
+ */
+public class StateTypeAdapter extends TypeAdapter<State> {
```
Maybe add `@NonNullByDefault`?

cpmeister commented on bundles/org.openhab.persistence.mongodb/src/main/java/org/openhab/persistence/mongodb/internal/MongoDBPersistenceService.java:
```diff
@@ -0,0 +1,390 @@
+        return item;
+    }
+
+    public static class MongoDBItem implements HistoricItem {
```
Maybe put this class into its own file for consistency

cpmeister commented on bundles/org.openhab.persistence.mysql/src/main/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java:
```diff
@@ -0,0 +1,743 @@
+
+            if (!matcher.group(1).equals("sqltype")) {
+                continue;
+            }
```
seems kinds inefficient

cpmeister commented on bundles/org.openhab.persistence.mysql/src/main/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java:
```diff
@@ -0,0 +1,743 @@
+        try {
+            sqlCmd = new String("INSERT INTO Items (ItemName) VALUES (?)");
+
+            statement = connection.prepareStatement(sqlCmd, Statement.RETURN_GENERATED_KEYS);
```
Statements are AutoClosable so put this in a try-with-resources

cpmeister commented on bundles/org.openhab.persistence.mysql/src/main/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java:
```diff
@@ -0,0 +1,743 @@
+
+            logger.debug("mySQL: Table created for item '{}' with datatype {} in SQL database.", itemName, mysqlType);
+            sqlTables.put(itemName, tableName);
+        } catch (Exception e) {
```
can this catch specific exceptions?

cpmeister commented on bundles/org.openhab.persistence.mysql/src/main/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java:
```diff
@@ -0,0 +1,743 @@
+            logger.debug("mySQL: query: {}", sqlCmd);
+
+            try {
+                statement = connection.prepareStatement(sqlCmd);
```
user try-with-resources

cpmeister commented on bundles/org.openhab.persistence.mysql/src/main/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java:
```diff
@@ -0,0 +1,743 @@
+
+            Statement st = connection.createStatement();
+            int result = st.executeUpdate("SHOW TABLES LIKE 'Items'");
+            st.close();
```
user try-with-resources

cpmeister replied:
```suggestion
            int result;
            try(Statement st = connection.createStatement()){
                result = st.executeUpdate("SHOW TABLES LIKE 'Items'");
            }
```

cpmeister commented on bundles/org.openhab.persistence.mysql/src/main/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java:
```diff
@@ -0,0 +1,743 @@
+
+            if (waitTimeout != -1) {
+                logger.debug("mySQL: Setting wait_timeout to {} seconds.", waitTimeout);
+                st = connection.createStatement();
```
same as above, there a bunch more below as well.

cpmeister commented on bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jItem.java:
```diff
@@ -0,0 +1,59 @@
+        return DateFormat.getDateTimeInstance().format(timestamp) + ": " + name + " -> " + state.toString();
+    }
+
+}
```
newline

cpmeister commented on bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jItem.java:
```diff
@@ -0,0 +1,59 @@
+
+    final private String name;
+    final private State state;
+    final private Date timestamp;
```
```suggestion
    private final String name;
    private final State state;
    private final Date timestamp;
```

cpmeister commented on bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java:
```diff
@@ -0,0 +1,590 @@
+                    }
+                    job = scheduler.schedule(() -> {
+                        store(item, name);
+                    }, 1, TimeUnit.SECONDS);
```
```suggestion
                    job = scheduler.schedule(() -> store(item, name), 1, TimeUnit.SECONDS);
```

cpmeister commented on bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/charts/RRD4jChartServlet.java:
```diff
@@ -0,0 +1,300 @@
+
+            return bi;
+        } catch (IOException e) {
+            logger.error("Error generating graph.", e);
```
Is this logging level appropriate?

cpmeister commented on features/openhab-addons-external/src/main/resources/conf/dynamodb.cfg:
```diff
@@ -0,0 +1,51 @@
+#writeCapacityUnits=1
+
+# table prefix used in the name of created tables
+#tablePrefix=openhab-
```
newline

cpmeister commented on features/openhab-addons-external/src/main/resources/conf/influxdb.cfg:
```diff
@@ -0,0 +1,14 @@
+
+# The name of the database, e.g. openhab.
+# Defaults to: openhab
+# db=<database>
```
newline

cpmeister commented on features/openhab-addons-external/src/main/resources/conf/openhabcloud.cfg:
```diff
@@ -1,19 +0,0 @@
-############################## openHAB Cloud Connector #############################
```
why is this class removed?

kaikreuzer replied:
This is part of the 2.5.x addons, which is auto-ported to 3.0.

cpmeister commented on features/openhab-addons-external/src/main/resources/conf/voicerss.cfg:
```diff
@@ -1,5 +0,0 @@
-
-# API key you got when you subscribed to the Voice RSS service
-# Open your profile to get the value: http://www.voicerss.org/personel/
-#apiKey=1234567890
```
why is this class removed?

kaikreuzer replied:
This is part of the 2.5.x addons, which is auto-ported to 3.0.

cpmeister commented on features/openhab-addons/pom.xml:
```diff
@@ -10,7 +10,7 @@
   </parent>
 
-  <artifactId>org.openhab.addons.features.karaf.openhab-addons</artifactId>
+  <artifactId>org.openhab.addons.features.karaf.openhab-addons3</artifactId>
```
why change this?

kaikreuzer replied:
To avoid the name clash between the auto-ported 2.5.x feature and the new one in 3.0.

cpmeister commented:
@kaikreuzer I guess I should have read the entirety of you post before I dove in. Looks like you can disregard most of my review LOL.

ssalonen commented on bundles/org.openhab.persistence.dynamodb/README.md:
```diff
@@ -0,0 +1,161 @@
+| writeCapacityUnits         | 1          |    No    | write capacity for the created tables                                                              |
+| tablePrefix                | `openhab-` |    No    | table prefix used in the name of created tables                                                    |
+| bufferCommitIntervalMillis | 1000       |    No    | Interval to commit (write) buffered data. In milliseconds.                                         |
+| bufferSize                 | 1000       |    No    | Internal buffer size which is used to batch writes to DynamoDB every `bufferCommitIntervalMillis`. |
```
it is not in kB, it's number of datapoints cc @kaikreuzer @cpmeister 

https://github.com/openhab/openhab-addons/blob/4d8226860e434a1987b58abeabc52ad68ea5fa22/bundles/org.openhab.persistence.dynamodb/src/main/java/org/openhab/persistence/dynamodb/internal/DynamoDBPersistenceService.java#L214
https://github.com/openhab/openhab-addons/blob/4d8226860e434a1987b58abeabc52ad68ea5fa22/bundles/org.openhab.persistence.dynamodb/src/main/java/org/openhab/persistence/dynamodb/internal/AbstractBufferedPersistenceService.java#L49

kaikreuzer replied:
Thanks @ssalonen, fixed with https://github.com/openhab/openhab-addons/pull/5275/commits/3ac6f3e4714db3a16dc178aa2c55037de4a9f3cc.

kaikreuzer commented on bundles/org.openhab.persistence.jdbc/src/main/java/org/openhab/persistence/jdbc/db/JdbcBaseDAO.java:
```diff
@@ -0,0 +1,583 @@
+        // databaseProps.setProperty("connectionTimeout",""+connectionTimeout);
+        // databaseProps.setProperty("idleTimeout", ""+idleTimeout);
+        // databaseProps.setProperty("maxLifetime", ""+maxLifetime);
+        // databaseProps.setProperty("validationTimeout",""+validationTimeout);
```
I guess this is there so that implementing classes can easily copy, past and adapt it - I'll leave it.

kaikreuzer commented:
Thanks for your review, @cpmeister!
I have now addressed (almost) all your comments, at least where I dared to touch the code.

cpmeister commented on bundles/org.openhab.persistence.mongodb/src/main/java/org/openhab/persistence/mongodb/internal/MongoDBItem.java:
```diff
@@ -0,0 +1,60 @@
+        return DateFormat.getDateTimeInstance().format(timestamp) + ": " + name + " -> " + state.toString();
+    }
+
+}
```
newline

cpmeister commented on bundles/org.openhab.persistence.mongodb/src/main/java/org/openhab/persistence/mongodb/internal/MongoDBItem.java:
```diff
@@ -0,0 +1,60 @@
+
+    final private String name;
+    final private State state;
+    final private Date timestamp;
```
```suggestion
    private final String name;
    private final State state;
    private final Date timestamp;
```

cpmeister commented on bundles/org.openhab.persistence.jdbc/src/main/java/org/openhab/persistence/jdbc/db/JdbcMysqlDAO.java:
```diff
@@ -0,0 +1,113 @@
+     * public Getters and Setters *
+     ******************************/
+
+}
```
newline

cpmeister commented on bundles/org.openhab.persistence.jdbc/src/main/java/org/openhab/persistence/jdbc/model/JdbcItem.java:
```diff
@@ -0,0 +1,65 @@
+        return builder.toString();
+    }
+
+}
```
newline

cpmeister commented on bundles/org.openhab.persistence.mysql/src/main/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java:
```diff
@@ -0,0 +1,717 @@
+    protected @NonNullByDefault({}) ItemRegistry itemRegistry;
+
+    // Error counter - used to reconnect to database on error
+    private int errCnt;
```
I don't know what kind of multithreading would go on in this class, but for threadsafety I would change this to AtomicInteger

cpmeister commented on bundles/org.openhab.persistence.mysql/src/main/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java:
```diff
@@ -0,0 +1,717 @@
+                logger.debug("mySQL: Setting wait_timeout to {} seconds.", waitTimeout);
+                st = connection.createStatement();
+                st.executeUpdate("SET SESSION wait_timeout=" + waitTimeout);
+                st.close();
```
```suggestion
                try(Statement st = connection.createStatement()){
                    st.executeUpdate("SET SESSION wait_timeout=" + waitTimeout);
                }
```

cpmeister commented on bundles/org.openhab.persistence.mysql/src/main/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java:
```diff
@@ -0,0 +1,717 @@
+                st.executeUpdate(
+                        "CREATE TABLE Items (ItemId INT NOT NULL AUTO_INCREMENT,ItemName VARCHAR(200) NOT NULL,PRIMARY KEY (ItemId));",
+                        Statement.RETURN_GENERATED_KEYS);
+                st.close();
```
```suggestion
                try(Statement st = connection.createStatement()){
                st.executeUpdate(
                        "CREATE TABLE Items (ItemId INT NOT NULL AUTO_INCREMENT,ItemName VARCHAR(200) NOT NULL,PRIMARY KEY (ItemId));",
                        Statement.RETURN_GENERATED_KEYS);
                }
```

cpmeister commented on bundles/org.openhab.persistence.mysql/src/main/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java:
```diff
@@ -0,0 +1,717 @@
+                sqlTables.put(rs.getString(2), "Item" + rs.getInt(1));
+            }
+            rs.close();
+            st.close();
```
```suggestion
            try(Statement st = connection.createStatement()){

            // Turn use of the cursor on.
            st.setFetchSize(50);
            try(ResultSet rs = st.executeQuery("SELECT ItemId, ItemName FROM Items")){
            while (rs.next()) {
                sqlTables.put(rs.getString(2), "Item" + rs.getInt(1));
            }
            }
            }
```

kaikreuzer commented:
@cpmeister Thanks, I addressed your latest comments.

kaikreuzer commented:
@ssalonen, @lewie and others: Now that those services are migrated and available in the 3.0-SNAPSHOT distro, we absolutely count on your to check whether your persistence services are functional - see my comment above (https://github.com/openhab/openhab-addons/pull/5275#issuecomment-598410298). Would be great, if you could go for this asap to make 3.0 a true replacement for 2.x!

lewie commented:
@kaikreuzer, 
very cool, an important step!

My detailed test scripts are written in Javascript and as far as I know there is currently no rule binding that can handle them for openHAB 3. Or is there? Now I tested the general installation and basic things like reading, writing and historical functions. The only thing is the CallType, which did not work, but it is outdated anyway?!

I have tested now, all JDBC Binding variants.

Tested persistence add-ons:
| Title | Sign | Installation | read/write |
| --- | --- | --- | --- |
| DynamoDB | dynamodb | - | - |
| InfluxSB | influxdb | - | - |
| JDBC H2 | jdbc-h2 | ok | ok |
| JDBC HSQLDB | jdbc-hsqldb | ok | ok |
| JDBC MariaDB | jdbc-mariadb | ok | ok |
| JDBC MySQL | jdbc-mysql | ok | ok |
| JDBC PostgreSQL | jdbc-postgresql | ok | ok |
| JDBC SQLite | jdbc-sqlite | ok | ok |
| JPA | jpa | - | - |
| MapDB | mapdb | - | - |
| MongoDB | mongodb | - | - |
| MySQL | mysql | - | - |

So, all tests I have done, turned out as expected, as in openHAB 2, in fact everything worked! 👍 💯 points!
Thank you very much, Kai for porting this!

One more remark. The second mysql binding (also JDBC) is redundant. At that time I built the JDBC binding for derby, h2, hsqldb, mariadb, mysql, postgres, sqlite, this was based on the original mysql binding. It was done just to reduce the redundancy of the same code for all JDBC based Drivers for databases. Even the database structure is fully compatible. Since the second obsolete mysql binding has no advantages, I suggest to finally let it die in openHAB 3 and recommend the JDBC-MySQL binding to users. What do you think about this saving of redundancy to maintain the code?

kaikreuzer commented:
Great, thank you very much for testing, the result is much better than I had hoped!

> written in Javascript and as far as I know there is currently no rule binding that can handle them for openHAB 3

Good question, I am actually not fully up to date on the status. But I am just working on removing the old rule engine and making the NGRE be installed by default and I think once we are there, it should not take long to have JS scripting back as well.

> The only thing is the CallType, which did not work, but it is outdated anyway?!

It is correct that the CallType does not exist anymore. But the CallItem is still there, it now only uses the StringListType internally. So it would be important to test this item type as well (and adapt the code to handle the StringListType correctly, which I indeed haven't done).

> What do you think about this saving of redundancy to maintain the code?

Yes, if the JDBC-MySQL service works as well as the other, I absolutely agree that we should remove the specific one! I'll create a PR immediately...

kaikreuzer commented:
... done: https://github.com/openhab/openhab-addons/pull/7306

ssalonen commented:
> It is correct that the CallType does not exist anymore. But the CallItem is still there, it now only uses the StringListType internally. So it would be important to test this item type as well (and adapt the code to handle the StringListType correctly, which I indeed haven't done).

The dynamodb integration tests failed with the CallItems due to this due to sorting differences, which is again due to changed serialization format (dynamodb persistence uses state.tostring). But this made me think..

Do we need to handle this in persistence bindings ? Otherwise old data is not readable. 

Are there other breaking changes in State.toString with other item types?

kaikreuzer commented:
> Do we need to handle this in persistence bindings ? Otherwise old data is not readable.

Pfew, good question. I actually doubt that anybody ever wants to read old CallItem states back, so I would say, we can document this as a breaking change and be fine with that. It is really a very small edge case and imho not worth to carry migration code in the add-on with us.

> Are there other breaking changes in State.toString with other item types?

No, not that I am aware of any.

lewie commented:
I am really happy, CallType works with string as value without errors! The data is stored as chars in the databases anyway. 👍 

> Great, thank you very much for testing, the result is much better than I had hoped!

...not just you!

> ...it should not take long to have JS scripting back as well.

I can hardly wait!

> Yes, if the JDBC-MySQL service works as well as the other, I absolutely agree that we should remove the specific one! I'll create a PR immediately...

Thanks, I think it's a good time for the deletion! And yes, there is also a point in the documentation about this: [migration-from-mysql-to-jdbc-persistence-services](https://www.openhab.org/addons/persistence/jdbc/#migration-from-mysql-to-jdbc-persistence-services)

Fragmentarily, I have followed the discussions the last Month. But now first time I have dealt with the current state of openHAB 3, great work, compliments to you all!
I am extremely excited about the new parts like the speed of the surface, that is contemporary!

ssalonen commented:
@kaikreuzer @lewie  how to configure the persistence addon in 3.0? Finally got the IDE resolved and the demo app launched but I do not see any obvious way to configure the addon. There is no ` runtime/conf/services` folder anymore.

Trying to test out the dynamodb with the full setup.

kaikreuzer commented:
> There is no runtime/conf/services folder anymore.

There has never been. the folder is `conf/services` and it is unchanged from openHAB 2. So the configuration is just done as before, no changes to that.

digitaldan commented:
>configuration is just done as before, no changes to that.

Trying use our new UI in a config file free use case made me realize we do not have the ability to configure persistence services through our API's, is this correct?  Do you think this will be needed for 3.0 ?

ssalonen commented:
I will try to place the file to old place and report back.

I think we used to have PaperUi UI for dynamodb persistence configuration but not anymore with 3.0 admin UI?

ghys commented:
> Is there admin UI for configuring the persistence addon similar to PaperUI? 

What does it look like in Paper UI? I have never seen it.

ssalonen commented:
@ghys 
This is how it looks like:
![Screenshot from 2020-05-10 19-49-12](https://user-images.githubusercontent.com/315866/81505389-838f3780-92f7-11ea-97aa-153f0ef5bbd4.png)

ghys commented:
So that's a service. It _should_ then appear under "other services" in the main settings menu.

ssalonen commented:
thanks @ghys. Unfortunately I do not see it there, only "Rule voice interpreter" and "Basic UI"

I have added the dynamodb persistence bundle in Run requirements of the demo app, nothing else.

As mentioned earlier, I'm not sure whether the bundle is loaded or not, since the shell does not seem to have the usual commands to diagnose running bundles...

ghys commented:
I've done the same, installed the DynamoDB add-on, and the configurable service should appear in the call to `/rest/services` but I don't see it...

![image](https://user-images.githubusercontent.com/2004147/81505573-876b8b80-92f0-11ea-890a-4c4c0861513b.png)

I'd be interesting to check if it's the case in 2.5.x so we can pinpoint what the actual issue is.

ghys commented:
Gave it a try on the latest 2.5.5 snapshot, and the service does indeed appear:

![image](https://user-images.githubusercontent.com/2004147/81505663-390abc80-92f1-11ea-99f8-e7fe5c81a8fe.png)

So there's something wrong in 3.0 wrt to these services, when it's fixed it will appear in the UI.

wborn commented:
The ESH-INF dirs should be renamed to OH-INF and they should be in `src/main/resources`:

```
$ find . | grep ESH-INF 
./bundles/org.openhab.persistence.influxdb/src/main/resources/ESH-INF
./bundles/org.openhab.persistence.influxdb/src/main/resources/ESH-INF/config
./bundles/org.openhab.persistence.influxdb/src/main/resources/ESH-INF/config/config.xml
./bundles/org.openhab.persistence.dynamodb/ESH-INF
./bundles/org.openhab.persistence.dynamodb/ESH-INF/config
./bundles/org.openhab.persistence.dynamodb/ESH-INF/config/config.xml

```

I'll test if that fixes it and create a PR for it.

kaikreuzer commented:
Oops, seems if I had missed those when migrating, sorry...
Thanks for fixing it. @wborn!

ssalonen commented:
Thank you all. Can confirm that UI & file works, success with persistence store and query.

Posted #7598 for some small improvements (mainly related to java8 -> java11)

cweitkamp commented on features/pom.xml:
```diff
@@ -16,6 +16,17 @@
   <name>openHAB Add-ons :: Features :: Karaf</name>
 
+  <properties>
+    <!-- JDBC database driver versions -->
```
@kaikreuzer Can you briefly if there was a special reason for adding the JDBC version properties in this file? It created confusion in https://github.com/openhab/openhab-addons/pull/9810#issuecomment-760533170.

Can we remove them? Or add them in the in the feature file of the bundle?

kaikreuzer replied:
I don't recall, sorry. If it works without it, we can remove them.

cweitkamp replied:
The code compiles successfully when I remove the properties from this file. I did not test installing the features.
